app/controllers/openid_connect/authorization_controller.rb

```diff
@@ -42,10 +39,22 @@ def check_sp_handoff_bounced
       true
     end
 
+    def confirm_user_has_aal3_mfa_if_requested
+      return unless aal3_policy.aal3_required?
+      return if mfa_policy.aal3_mfa_enabled?
+      # session[:needs_to_setup_piv_cac_after_sign_in] = true
+      redirect_to two_factor_options_url
+    end
+
     def confirm_user_is_authenticated_with_fresh_mfa
       bump_auth_count unless user_fully_authenticated?
       return confirm_two_factor_authenticated(request_id) unless user_fully_authenticated?
-      redirect_to user_two_factor_authentication_url if remember_device_expired_for_sp?
+      redirect_to user_two_factor_authentication_url if device_not_remembered?
+    end
+
+    def device_not_remembered?
+      remember_device_expired_for_sp? ||
+        (aal3_policy.aal3_required? && mfa_policy.aal3_mfa_enabled? && !aal3_policy.aal3_used?)
```

Comment:
yeah this is that same triplet of predicates....I think these should definitely be a method together. Maybe if not on the controller, there's some sort of `AalContext` that is session + user policy we could use?

```
aal_context.aal3_required_and_enabled_but_not_used?
```